omegaconf/basecontainer.py

```diff
@@ -257,6 +262,34 @@ def convert(val: Node) -> Any:
 
         assert False
 
+    def _instantiate_structured_config_impl(self, instance_data: Dict[str, Any]) -> Any:
+        """Instantiate an instance of `self._metadata.object_type`, populated by `instance_data`."""
+        from ._utils import get_structured_config_field_names
+
+        object_type = self._metadata.object_type
+        object_type_field_names = set(get_structured_config_field_names(object_type))
+
+        field_items: Dict[str, Any] = {}
+        nonfield_items: Dict[str, Any] = {}
+        for k, v in instance_data.items():
+            if _is_missing_literal(v):
+                self._format_and_raise(
+                    key=k,
+                    value=None,
+                    cause=MissingMandatoryValue(
+                        "Structured config of type `$OBJECT_TYPE` has missing mandatory value: $KEY"
+                    ),
+                )
+            if k in object_type_field_names:
+                field_items[k] = v
+            else:
+                nonfield_items[k] = v
+
+        result = object_type(**field_items)
+        for k, v in nonfield_items.items():
+            setattr(result, k, v)
+        return result
```

Comment:
I realized that this would not work for `ClassExtendingDictIntKeys(Dict[int, VT])`.

I wonder if we should only support extending `Dict[str, VT]` in general, eliminating this problem and simplifying a handful of other cases which are ill defined.

Reply:
Let's put down the motivation initial motivations help guide the decision:
1. DictConfig now supports non string keys.
2. Extending a Dict in a Structured config (dataclass/attr class) is used (one can say abused) as a way to flag that node as open to additions (either via setattr or via setitem).

The motivation for the second case only makes sense for string keys.
You may remember I brought up the case for extending Dict when you added support for additional key types.
I am wondering now if we should only support extending Dict[str, VT].
`Dict[Any, VT]` can issue a deprecation warning saying only `Dict[str, VT]` is supported, and it can treat it like `Dict[str, VT]`.

At that point, the problem I am bringing up here will disappear.
If that plan makes sense to you, it should be in a followup PR, not this one.

Reply:
> At that point, the problem I am bringing up here will disappear.
> If that plan makes sense to you, it should be in a followup PR, not this one.

Fine with me.

Reply:
I've created issue https://github.com/omry/omegaconf/issues/657 to keep track of this.